CDAP-6927 Move listing of streams to StreamHandler

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-6927

Build : http://builds.cask.co/browse/CDAP-DUT4632
http://builds.cask.co/browse/CDAP-RUT10

gokulavasan committed:
CDAP-6927 Move listing of streams to StreamHandler and also add authorization

Adding unit test as part of the PR comment

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java:
```diff
@@ -1,80 +0,0 @@
-   * Returns a list of streams in a namespace.
-   */
-  @GET
-  @Path("/streams")
```
there aren't any compatibility implications of moving this API to another service, right?

gokulavasan replied:
not that I know of. @chtyim Do you have any concerns?

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java:
```diff
@@ -177,6 +182,18 @@ public void destroy(HandlerContext context) {
+  public void listStreams(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId) throws Exception {
+    List<StreamSpecification> specificationList = streamAdmin.listStreams(new NamespaceId(namespaceId));
+    List<StreamDetail> streamDetails = Lists.newArrayListWithCapacity(specificationList.size());
```
just use new ArrayList<>(specifications.size())?

anwar6953 commented:
Where is the authorization being added? Is it implicit?

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java:
```diff
@@ -177,6 +182,18 @@ public void destroy(HandlerContext context) {
+  @Path("/")
+  public void listStreams(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId) throws Exception {
+    List<StreamSpecification> specificationList = streamAdmin.listStreams(new NamespaceId(namespaceId));
```
nit: call it specifications, since the datatype already suggests that its a list?

bdmogal commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:
```diff
@@ -111,13 +111,7 @@ private String getV3RoutingService(String [] uriParts, AllowedMethod requestMeth
       requestMethod.equals(AllowedMethod.GET)) {
       return Constants.Service.APP_FABRIC_HTTP;
     } else if ((uriParts.length >= 4) && uriParts[1].equals("namespaces") && uriParts[3].equals("streams")) {
-      // /v3/namespaces/<namespace>/streams goes to AppFabricHttp
```
aren't there any tests in RouterPathTest that need to be updated with this change?

bdmogal replied:
nvm. saw them

bdmogal commented on cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTest.java:
```diff
@@ -205,6 +208,22 @@ public void testSimpleStreamEnqueue() throws Exception {
+  public void testListStreams() throws Exception {
+    List<StreamDetail> specs = listStreams(NamespaceId.DEFAULT);
+    Assert.assertTrue(specs.isEmpty());
+    StreamId s1 = new StreamId(NamespaceId.DEFAULT.getNamespace(), "stream1");
```
just NamespaceId.DEFAULT.stream("stream1")?

bdmogal commented on cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTest.java:
```diff
@@ -205,6 +208,22 @@ public void testSimpleStreamEnqueue() throws Exception {
+    List<StreamDetail> specs = listStreams(NamespaceId.DEFAULT);
+    Assert.assertTrue(specs.isEmpty());
+    StreamId s1 = new StreamId(NamespaceId.DEFAULT.getNamespace(), "stream1");
+    StreamId s2 = new StreamId(NamespaceId.DEFAULT.getNamespace(), "stream2");
```
ditto

bdmogal commented:
@gokulavasan just a few minor comments. Should we add tests with authorization enabled?

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java:
```diff
@@ -275,6 +275,12 @@ public void upgrade() throws Exception {
 
   @Override
+  public List<StreamSpecification> listStreams(NamespaceId namespaceId) throws Exception {
+    ensureAccess(namespaceId);
```
@gokulavasan this seems wrong. this should instead filter the streams in the namespace and only return the ones to which the user has access.

gokulavasan replied:
@bdmogal Ah sorry. I was misled by this 'list   READ/WRITE/ADMIN/ALL (Namespace)' in the wiki. I will do the filtering.

bdmogal replied:
👍 please update the wiki as well

gokulavasan commented:
@bdmogal Addressed comments. Please take a look. Can you point me to any example where we unit test HTTP handlers with authorization handlers turned on?

bdmogal commented:
Look at DatasetServiceAuthorizationTest

gokulavasan commented:
@bdmogal Added a test case to test listing of streams in StreamAdminTest. Please check the last commit. Thanks!

bdmogal commented:
LGTM

gokulavasan commented:
Thanks for the review @bdmogal ! I have started another build under CDAP-RUT. Will merge after that passes.

gokulavasan commented:
Build passed. Merging.
